RationaleMCP/0031/differences.md

```diff
@@ -341,3 +341,313 @@ end 'LineB';
 ```
 
 If the `LineA` variant ends up being valid in full Modelica, then this is the form that will also be used for Flat Modelica.  Otherwise, Flat Modelica will use the `LineB` form.
+
+### Final modification
+
+The concept of being final in full Modelica implies two different things:
+- Further modification is not possible.  This can be verified in the reduction from full Modelica to Flat Modelica, and there is no real need to express this constraint also in the Flat Modelica model.  (It could be useful for expressing constraints for hand-written Flat Modelica, but it is a language feature we could add later if requested.)
+- Parameter values and `start` attributes cannot be modified after translation.  This is something that can't just be verified during the reduction from full Modelica to Flat Modelica.  Instead, final parameter declaration equations are turned into initial equations, and then the same technique is used to handle final modification of `start`.  Details of this are given the sections below on initialization of parameters and time-varying variables.
+
+
+## Initialization of parameters
+
+In Flat Modelica, a parameter's declaration equation shall be solved with respect to the parameter, allowing the right hand side to be overridden during initialization (that is, after translation).  This is similar to a full Modelica non-final parameter with `fixed = true`.
+
+For example, the full Modelica
+```
+parameter Real p(fixed = true) = 4.2;
+```
+translates to the Flat Modelica
+```
+parameter Real 'p' = 4.2; /* Presence of declaration equation corresponds to full Modelica fixed = true. */
+```
+
+In Flat Modelica, a parameter without declaration equation shall be solvable from equations given in the `initial equation` section.  This corresponds directly to the full Modelica parameters with `fixed = false`.
+
+For example, the full Modelica
+```
+  parameter Real p(fixed = false);
+initial equation
+  p^2 + p = 1;
+```
+translates to the Flat Modelica
+```
+  parameter Real 'p'; /* Full Modelica parameter with fixed = false. */
+initial equation
+  'p'^2 + 'p' = 1;
+```
+
+The same mechanism is also able to represent a full Modelica final declaration equation.
+
+For example, the full Modelica
```

Comment:
I don't see that replacing final with an initial equation is helpful. We lose important context, and the only simplification is removing a keyword; that is used by many other languages. If we really dislike "final" we could have all "final" parameters as "protected" or some other mechanism that prevent changes.

Reply:
The `final` has an important role to play in full Modelica as well as in other languages.  In Flat Modelica, on the other hand, we've noted that `final` has mostly lost its meaning since it is primarily a guard against the modifications that take place during instantiation.  The aspect of final that we are now working to get rid of is something different and specific to the parameter variability that doesn't have a counterpart in the mainstream general purpose languages where `final` is a useful concept: protection against certain equations (parameter bindings) from being modified after translation.  Based on this there are good reasons to get rid of this other aspect of `final` in Flat Modelica:
- The `final` concept becomes misleading since it is no longer at all related to protection against translation-time modifications.
- That equations can't be modified after translation is a well understood principle in Modelica, and would serve us well for modeling the rule that a `final` parameter must not be modified after translation.
- Getting rid of things we don't really need is in the spirit of Flat Modelica.

Reply:
To me keeping final remains good, since it protects against modification of parts of the class with minimal impact, although in some cases an alternative would be to use protected.

As an example: one important aspect of C++ classes is that public, protected, private only influence the access and nothing else (well, a bit technical for standard-layout classes). That means that you can take an existing class and modify it to grant access (for testing) or deny access (if nothing external uses it) knowing that it will remain working after this  change.

That creates a separation of concern that is important.

Having to move final bindings to another section doesn't respect this separation of concerns.

Reply:
The things with different visibility in Modelica are the component declarations; equations don't have visibility in this sense.  Moving a declaration equation to an equation section doesn't change the visibility of the component.

Reply:
> The things with different visibility in Modelica are the component declarations; equations don't have visibility in this sense. Moving a declaration equation to an equation section doesn't change the visibility of the component.

It wasn't implying that it was exactly the same, but that changing a binding from final to non-final (or vice-versa) in current Modelica satisfies the same requirement of **separation of concern.**

Meaning that adding/removing final is a localized change that does not impact the rest of the semantics, whereas the proposal doesn't satisfy this and requires that the equation is moved from being attached to a declaration (or inside it) to an equation section - with no obvious way of reversing it.

In Modelica we also have something intermediate - parameters that are disabled and/or hidden in "Advanced" tabs. The goal from a user-perspective is to prevent modification (similarly as final) - but with possible exceptions; with minimal changes of the model code - since the goal is to ensure that the model behaves the same as before.